apollo-router/src/apollo_telemetry.rs

```diff
@@ -0,0 +1,553 @@
+//! [`Span`]: crate::trace::Span
+//! [`Report`]: apollo_spaceport::report::Report
+//! [`Reporter`]: apollo_spaceport::Reporter
+//!
+//! # Examples
+//!
+//! ```ignore
+//! use crate::apollo_telemetry;
+//! use opentelemetry::trace::Tracer;
+//! use opentelemetry::global::shutdown_tracer_provider;
+//!
+//! fn main() {
+//!     let tracer = apollo_telemetry::new_pipeline()
+//!         .install_simple();
+//!
+//!     tracer.in_span("doing_work", |cx| {
+//!         // Traced app logic here...
+//!     });
+//!
+//!     shutdown_tracer_provider(); // sending remaining spans
+//! }
+//! ```
+use apollo_parser::{ast, Parser};
+use apollo_spaceport::report::{ContextualizedStats, QueryLatencyStats, StatsContext};
+use apollo_spaceport::{Reporter, ReporterGraph};
+use async_trait::async_trait;
+use once_cell::sync::OnceCell;
+use opentelemetry::{
+    global,
+    runtime::Tokio,
+    sdk,
+    sdk::export::{
+        trace::{ExportResult, SpanData, SpanExporter},
+        ExportError,
+    },
+    trace::{TraceError, TracerProvider},
+    Value,
+};
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::time::Duration;
+use tokio::task::JoinError;
+
+use crate::configuration::{SpaceportConfig, StudioGraph};
+
+pub(crate) const DEFAULT_SERVER_URL: &str = "https://127.0.0.0:50051";
+pub(crate) const DEFAULT_LISTEN: &str = "0.0.0.0:50051";
+
+/// Pipeline builder
+#[derive(Debug)]
+pub struct PipelineBuilder {
+    graph_config: Option<StudioGraph>,
+    spaceport_config: Option<SpaceportConfig>,
+    trace_config: Option<sdk::trace::Config>,
+}
+
+/// Create a new apollo telemetry exporter pipeline builder.
+pub fn new_pipeline() -> PipelineBuilder {
+    PipelineBuilder::default()
+}
+
+impl Default for PipelineBuilder {
+    /// Return the default pipeline builder.
+    fn default() -> Self {
+        Self {
+            graph_config: None,
+            spaceport_config: None,
+            trace_config: None,
+        }
+    }
+}
+
+impl PipelineBuilder {
+    /// Assign the SDK trace configuration.
+    #[allow(dead_code)]
+    pub fn with_trace_config(mut self, config: sdk::trace::Config) -> Self {
+        self.trace_config = Some(config);
+        self
+    }
+
+    /// Assign graph identification configuration
+    pub fn with_graph_config(mut self, config: &Option<StudioGraph>) -> Self {
+        self.graph_config = config.clone();
+        self
+    }
+
+    /// Assign spaceport reporting configuration
+    pub fn with_spaceport_config(mut self, config: &Option<SpaceportConfig>) -> Self {
+        self.spaceport_config = config.clone();
+        self
+    }
+
+    /// Install the apollo telemetry exporter pipeline with the recommended defaults.
+    pub fn install_batch(mut self) -> Result<sdk::trace::Tracer, ApolloError> {
+        let exporter = self.get_exporter()?;
+
+        let batch = sdk::trace::BatchSpanProcessor::builder(exporter, Tokio)
+            .with_max_queue_size(65_536)
```

Comment:
could that be exposed through the configuration file? IIRC it's available for the other exporters. We should provide good defaults for that, but making those parameters accessible might be what makes production incidents more manageable

Reply:
Yeah. I'm not sure how best to deal with this. The OTEL stuff uses environment variables and we use them in trace.rs as well. So, instead of going with configuration perhaps we could embrace the environment variables but specify lower bounds and log a warning as appropriate.

i.e.: default max_queue_size is 65_536, if someone specified OTEL_BSP_MAX_QUEUE_SIZE < 65_536 we log a message saying min is 65_536. (likewise for batch)

The reason being that very small batch and queue sizes make opentelemetry fairly useless in production environments for apollo telemetry.

Reply:
yes, honoring the env variables should be ok. I'm not too fond of the approach on batching here, but we have to conform to rust-opentelemetry's design

Reply:
done: 042f8379ba0ac8b1b1e378cf21401f03e76d4531